src/linear_algebra/finite_dimensional.lean

```diff
@@ -580,45 +582,52 @@ end
 
 end
 
-end division_ring
-
-section field
-
-variables {K : Type u} {V : Type v} [field K] [add_comm_group V] [module K V]
-{V₂ : Type v'} [add_comm_group V₂] [module K V₂]
-
 /-- In a vector space with dimension 1, each set {v} is a basis for `v ≠ 0`. -/
+@[simps]
 noncomputable def basis_singleton (ι : Type*) [unique ι]
   (h : finrank K V = 1) (v : V) (hv : v ≠ 0) :
   basis ι K V :=
 let b := basis_unique ι h in
-b.map (linear_equiv.smul_of_unit (units.mk0
-  (b.repr v default)
-  (mt basis_unique.repr_eq_zero_iff.mp hv)))
+let h : b.repr v default ≠ 0 := mt basis_unique.repr_eq_zero_iff.mp hv in
+basis.of_repr
+{ to_fun := λ w, finsupp.single default (b.repr w default / b.repr v default),
+  inv_fun := λ f, f default • v,
+  map_add' := by simp [add_div],
+  map_smul' := by simp [mul_div],
+  left_inv := λ w, begin
+    apply_fun b.repr using b.repr.to_equiv.injective,
+    apply_fun equiv.finsupp_unique,
+    simp only [linear_equiv.map_smulₛₗ, finsupp.coe_smul, finsupp.single_eq_same, ring_hom.id_apply,
+      smul_eq_mul, pi.smul_apply, equiv.finsupp_unique_apply],
+    exact div_mul_cancel _ h,
+  end ,
+  right_inv := λ f, begin
+    ext,
+    simp only [linear_equiv.map_smulₛₗ, finsupp.coe_smul, finsupp.single_eq_same, ring_hom.id_apply,
+      smul_eq_mul, pi.smul_apply],
+    exact mul_div_cancel _ h,
+  end, }
```

Comment:
Is this indicating that we are missing some general API lower in the hierarchy? I would expect that we can keep this defn about as short as the old defn.

Reply:
I'm failing to see what could be extracted out of this... The short-cut that works in the commutative case just doesn't anymore.